File: pdu/python/std_msgs/pdu_conv_MultiArrayDimension.py

import struct
from .pdu_pytype_MultiArrayDimension import MultiArrayDimension
from ..pdu_utils import *
from .. import binary_io

# dependencies for the generated Python class



def pdu_to_py_MultiArrayDimension(binary_data: bytearray) -> MultiArrayDimension:
    py_obj = MultiArrayDimension()
    meta_parser = binary_io.PduMetaDataParser()
    meta = meta_parser.load_pdu_meta(binary_data)
    if meta is None:
        raise ValueError("Invalid PDU binary data: MetaData not found or corrupted")
    binary_read_recursive_MultiArrayDimension(meta, binary_data, py_obj, binary_io.PduMetaData.PDU_META_DATA_SIZE)
    return py_obj


def binary_read_recursive_MultiArrayDimension(meta: binary_io.PduMetaData, binary_data: bytearray, py_obj: MultiArrayDimension, base_off: int):
    # array_type: single 
    # data_type: primitive 
    # member_name: label 
    # type_name: string 
    # offset: 0 size: 128 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 0, 128)
    py_obj.label = binary_io.binTovalue("string", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: size 
    # type_name: uint32 
    # offset: 128 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 128, 4)
    py_obj.size = binary_io.binTovalue("uint32", bin)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: stride 
    # type_name: uint32 
    # offset: 132 size: 4 
    # array_len: 1

    
    bin = binary_io.readBinary(binary_data, base_off + 132, 4)
    py_obj.stride = binary_io.binTovalue("uint32", bin)
    
    return py_obj



def py_to_pdu_MultiArrayDimension(py_obj: MultiArrayDimension) -> bytearray:
    binary_data = bytearray()
    base_allocator = DynamicAllocator(False)
    bw_container = BinaryWriterContainer(binary_io.PduMetaData())
    binary_write_recursive_MultiArrayDimension(0, bw_container, base_allocator, py_obj)

    # メタデータの設定
    total_size = base_allocator.size() + bw_container.heap_allocator.size() + binary_io.PduMetaData.PDU_META_DATA_SIZE
    bw_container.meta.total_size = total_size
    bw_container.meta.heap_off = binary_io.PduMetaData.PDU_META_DATA_SIZE + base_allocator.size()

    # binary_data のサイズを total_size に調整
    if len(binary_data) < total_size:
        binary_data.extend(bytearray(total_size - len(binary_data)))
    elif len(binary_data) > total_size:
        del binary_data[total_size:]

    # メタデータをバッファにコピー
    binary_io.writeBinary(binary_data, 0, bw_container.meta.to_bytes())

    # 基本データをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.base_off, base_allocator.to_array())

    # ヒープデータをバッファにコピー
    binary_io.writeBinary(binary_data, bw_container.meta.heap_off, bw_container.heap_allocator.to_array())

    return binary_data

def binary_write_recursive_MultiArrayDimension(parent_off: int, bw_container: BinaryWriterContainer, allocator, py_obj: MultiArrayDimension):
    # array_type: single 
    # data_type: primitive 
    # member_name: label 
    # type_name: string 
    # offset: 0 size: 128 
    # array_len: 1
    type = "string"
    off = 0

    
    bin = binary_io.typeTobin(type, py_obj.label)
    bin = get_binary(type, bin, 128)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: size 
    # type_name: uint32 
    # offset: 128 size: 4 
    # array_len: 1
    type = "uint32"
    off = 128

    
    bin = binary_io.typeTobin(type, py_obj.size)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    
    # array_type: single 
    # data_type: primitive 
    # member_name: stride 
    # type_name: uint32 
    # offset: 132 size: 4 
    # array_len: 1
    type = "uint32"
    off = 132

    
    bin = binary_io.typeTobin(type, py_obj.stride)
    bin = get_binary(type, bin, 4)
    allocator.add(bin, expected_offset=parent_off + off)
    

File: pdu/javascript/std_msgs/pdu_conv_MultiArrayDimension.js
import * as PduUtils from '../pdu_utils.js';
import { MultiArrayDimension } from './pdu_jstype_MultiArrayDimension.js';


/**
 * Deserializes a binary PDU into a MultiArrayDimension object.
 * @param {ArrayBuffer} binary_data
 * @returns { MultiArrayDimension }
 */
export function pduToJs_MultiArrayDimension(binary_data) {
    const js_obj = new MultiArrayDimension();
    const meta_parser = new PduUtils.PduMetaDataParser();
    const meta = meta_parser.load_pdu_meta(binary_data);
    if (meta === null) {
        throw new Error("Invalid PDU binary data: MetaData not found or corrupted");
    }
    binary_read_recursive_MultiArrayDimension(meta, binary_data, js_obj, meta.base_off);
    return js_obj;
}

export function binary_read_recursive_MultiArrayDimension(meta, binary_data, js_obj, base_off) {
    const view = new DataView(binary_data);
    const littleEndian = true;
    // member: label, type: string (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 0, 128);
        js_obj.label = PduUtils.binToValue("string", bin);
    }
    
    // member: size, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 128, 4);
        js_obj.size = PduUtils.binToValue("uint32", bin);
    }
    
    // member: stride, type: uint32 (primitive)

    
    {
        const bin = PduUtils.readBinary(binary_data, base_off + 132, 4);
        js_obj.stride = PduUtils.binToValue("uint32", bin);
    }
    
    return js_obj;
}

/**
 * Serializes a MultiArrayDimension object into a binary PDU.
 * @param { MultiArrayDimension } js_obj
 * @returns {ArrayBuffer}
 */
export function jsToPdu_MultiArrayDimension(js_obj) {
    const base_allocator = new PduUtils.DynamicAllocator();
    const bw_container = new PduUtils.BinaryWriterContainer(new PduUtils.PduMetaData());

    binary_write_recursive_MultiArrayDimension(0, bw_container, base_allocator, js_obj);

    const base_data_size = base_allocator.size();
    const heap_data_size = bw_container.heap_allocator.size();
    
    bw_container.meta.heap_off = PduUtils.PDU_META_DATA_SIZE + base_data_size;
    bw_container.meta.total_size = bw_container.meta.heap_off + heap_data_size;

    const final_buffer = new ArrayBuffer(bw_container.meta.total_size);
    const final_view = new Uint8Array(final_buffer);

    PduUtils.writeBinary(final_view, 0, bw_container.meta.to_bytes());
    PduUtils.writeBinary(final_view, bw_container.meta.base_off, base_allocator.toArray());
    PduUtils.writeBinary(final_view, bw_container.meta.heap_off, bw_container.heap_allocator.toArray());

    return final_buffer;
}

export function binary_write_recursive_MultiArrayDimension(parent_off, bw_container, allocator, js_obj) {
    const littleEndian = true;
    // member: label, type: string (primitive)

    
    {
        const bin = PduUtils.typeToBin("string", js_obj.label, 128);
        allocator.add(bin, parent_off + 0);
    }
    
    // member: size, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.size, 4);
        allocator.add(bin, parent_off + 128);
    }
    
    // member: stride, type: uint32 (primitive)

    
    {
        const bin = PduUtils.typeToBin("uint32", js_obj.stride, 4);
        allocator.add(bin, parent_off + 132);
    }
    
}
